File: Backend/src/main/java/com/example/Backend/service/DeliveryService.java
package com.example.Backend.service;

import com.example.Backend.dto.DeliveryDTO;
import com.example.Backend.entity.Delivery;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.Map;
import java.util.UUID;

public interface DeliveryService {

    /**
     * Create delivery for order
     * 
     * @param orderId     Order ID
     * @param deliveryDTO Delivery data
     * @return Created delivery
     */
    Delivery createDelivery(UUID orderId, DeliveryDTO deliveryDTO);

    /**
     * Get delivery by ID
     * 
     * @param deliveryId Delivery ID
     * @return Delivery entity
     */
    Delivery getDeliveryById(UUID deliveryId);

    /**
     * Get delivery by order ID
     * 
     * @param orderId Order ID
     * @return Delivery entity
     */
    Delivery getDeliveryByOrderId(UUID orderId);

    /**
     * Get delivery by tracking number
     * 
     * @param trackingNumber ONdelivery tracking number
     * @return Delivery entity
     */
    Delivery getDeliveryByTrackingNumber(String trackingNumber);

    /**
     * Get all deliveries
     * 
     * @param pageable Pagination parameters
     * @return Page of deliveries
     */
    Page<Delivery> getAllDeliveries(Pageable pageable);

    /**
     * Get deliveries by status
     * 
     * @param status   Delivery status
     * @param pageable Pagination parameters
     * @return Page of deliveries
     */
    Page<Delivery> getDeliveriesByStatus(String status, Pageable pageable);

    /**
     * Get pending deliveries
     * 
     * @param pageable Pagination parameters
     * @return Page of pending deliveries
     */
    Page<Delivery> getPendingDeliveries(Pageable pageable);

    /**
     * Get active deliveries (in transit)
     * 
     * @param pageable Pagination parameters
     * @return Page of active deliveries
     */
    Page<Delivery> getActiveDeliveries(Pageable pageable);

    /**
     * Get deliveries by courier
     * 
     * @param courierName Courier name
     * @param pageable    Pagination parameters
     * @return Page of deliveries
     */
    Page<Delivery> getDeliveriesByCourier(String courierName, Pageable pageable);

    /**
     * Update delivery status
     * 
     * @param deliveryId Delivery ID
     * @param status     New status
     * @return Updated delivery
     */
    Delivery updateStatus(UUID deliveryId, String status);

    /**
     * Mark as picked up
     * 
     * @param deliveryId  Delivery ID
     * @param courierName Courier name
     * @return Updated delivery
     */
    Delivery markAsPickedUp(UUID deliveryId, String courierName);

    /**
     * Mark as in transit
     * 
     * @param deliveryId Delivery ID
     * @return Updated delivery
     */
    Delivery markAsInTransit(UUID deliveryId);

    /**
     * Mark as delivered
     * 
     * @param deliveryId Delivery ID
     * @return Updated delivery
     */
    Delivery markAsDelivered(UUID deliveryId);

    /**
     * Track delivery (sync with ONdelivery API)
     * 
     * @param trackingNumber Tracking number
     * @return Updated delivery with tracking info
     */
    Delivery trackDelivery(String trackingNumber);

    /**
     * Get delivery statistics
     * 
     * @return Map of counts by status
     */
    Map<String, Long> getDeliveryStatistics();

    /**
     * Get average delivery time
     * 
     * @return Average time in hours
     */
    Double getAverageDeliveryTime();

    /**
     * Create deliveries for existing shipped orders that don't have one
     * 
     * @return Number of deliveries created
     */
    int createDeliveriesForShippedOrders();

    /**
     * Get debug info about orders and deliveries
     * 
     * @return Map with debug information
     */
    Map<String, Object> getDebugInfo();

    /**
     * Sync tracking numbers from deliveries to orders
     * 
     * @return Number of orders synced
     */
    int syncTrackingNumbersToOrders();

    /**
     * Auto-create delivery from confirmed order
     * Called automatically when order status changes to CONFIRMED
     * 
     * @param order The confirmed order
     * @return Created delivery
     */
    Delivery createDeliveryFromOrder(com.example.Backend.entity.Order order);

    /**
     * Create deliveries for all confirmed orders that don't have one
     * 
     * @return Number of deliveries created
     */
    int createDeliveriesForConfirmedOrders();

    /**
     * Ensure simulation is running for active deliveries
     * Used when user opens tracking page - restarts simulation if needed
     * 
     * @param deliveryId The delivery ID
     */
    void ensureSimulationRunning(java.util.UUID deliveryId);
}

File: frontend-web/src/app/core/services/delivery.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';

export interface Delivery {
    id: string;
    trackingNumber: string;
    order: any;
    deliveryAddress: string;
    address: string;
    recipientName: string;
    recipientPhone: string;
    status: 'PROCESSING' | 'PICKED_UP' | 'IN_TRANSIT' | 'OUT_FOR_DELIVERY' | 'DELIVERED' | 'FAILED' | 'CANCELLED';
    courierName?: string;
    driverName?: string;
    driverPhone?: string;
    estimatedDelivery?: Date;
    estimatedDeliveryDate?: Date;
    actualDeliveryDate?: Date;
    actualDelivery?: Date;
    deliveryNotes?: string;
    notes?: string;
    currentLatitude?: number;
    currentLongitude?: number;
    createdAt: Date;
    updatedAt: Date;
}

export interface Page<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

export interface DeliveryStats {
    totalDeliveries: number;
    processing: number;
    inTransit: number;
    outForDelivery: number;
    delivered: number;
    failed: number;
    averageDeliveryTime: number;
    onTimeRate?: number;
}

// Backend response type (uses UPPER_CASE keys)
interface BackendDeliveryStats {
    TOTAL?: number;
    PROCESSING?: number;
    IN_TRANSIT?: number;
    OUT_FOR_DELIVERY?: number;
    DELIVERED?: number;
    FAILED?: number;
}

@Injectable({
    providedIn: 'root'
})
export class DeliveryService {
    private apiUrl = `${environment.apiUrl}/delivery`;

    constructor(private http: HttpClient) { }

    // ==================== DELIVERY CRUD ====================

    createDelivery(orderId: string, delivery: Partial<Delivery>): Observable<Delivery> {
        const params = new HttpParams().set('orderId', orderId);
        return this.http.post<Delivery>(this.apiUrl, delivery, { params });
    }

    getDeliveryById(id: string): Observable<Delivery> {
        return this.http.get<Delivery>(`${this.apiUrl}/${id}`);
    }

    getDeliveryByOrderId(orderId: string): Observable<Delivery> {
        return this.http.get<Delivery>(`${this.apiUrl}/order/${orderId}`);
    }

    trackDelivery(trackingNumber: string): Observable<Delivery> {
        return this.http.get<Delivery>(`${this.apiUrl}/track/${trackingNumber}`);
    }

    getDeliveryByTrackingNumber(trackingNumber: string): Observable<Delivery> {
        return this.http.get<Delivery>(`${this.apiUrl}/tracking/${trackingNumber}`);
    }

    // ==================== ADMIN QUERIES ====================

    getAllDeliveries(page: number = 0, size: number = 20, sort: string = 'createdAt'): Observable<Page<Delivery>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString())
            .set('sort', sort);
        return this.http.get<Page<Delivery>>(this.apiUrl, { params });
    }

    getDeliveriesByStatus(status: string, page: number = 0, size: number = 20): Observable<Page<Delivery>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Delivery>>(`${this.apiUrl}/status/${status}`, { params });
    }

    getPendingDeliveries(page: number = 0, size: number = 20): Observable<Page<Delivery>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Delivery>>(`${this.apiUrl}/pending`, { params });
    }

    getActiveDeliveries(page: number = 0, size: number = 20): Observable<Page<Delivery>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Delivery>>(`${this.apiUrl}/active`, { params });
    }

    getDeliveriesByCourier(courierName: string, page: number = 0, size: number = 20): Observable<Page<Delivery>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());
        return this.http.get<Page<Delivery>>(`${this.apiUrl}/courier/${courierName}`, { params });
    }

    // ==================== STATUS UPDATES ====================

    updateStatus(id: string, status: string): Observable<Delivery> {
        return this.http.patch<Delivery>(`${this.apiUrl}/${id}/status`, { status });
    }

    markAsPickedUp(id: string, courierName: string): Observable<Delivery> {
        return this.http.patch<Delivery>(`${this.apiUrl}/${id}/picked-up`, { courierName });
    }

    markAsInTransit(id: string): Observable<Delivery> {
        return this.http.patch<Delivery>(`${this.apiUrl}/${id}/in-transit`, {});
    }

    markAsDelivered(id: string): Observable<Delivery> {
        return this.http.patch<Delivery>(`${this.apiUrl}/${id}/delivered`, {});
    }

    // ==================== STATISTICS ====================

    getStatistics(): Observable<DeliveryStats> {
        return this.http.get<BackendDeliveryStats>(`${this.apiUrl}/statistics`).pipe(
            map(backendStats => ({
                totalDeliveries: backendStats.TOTAL || 0,
                processing: backendStats.PROCESSING || 0,
                inTransit: backendStats.IN_TRANSIT || 0,
                outForDelivery: backendStats.OUT_FOR_DELIVERY || 0,
                delivered: backendStats.DELIVERED || 0,
                failed: backendStats.FAILED || 0,
                averageDeliveryTime: 0, // Will be fetched separately
                onTimeRate: 0
            }))
        );
    }

    getAverageDeliveryTime(): Observable<number> {
        return this.http.get<number>(`${this.apiUrl}/average-time`);
    }

    /**
     * Sync confirmed orders - creates deliveries for confirmed orders that don't have one
     */
    syncConfirmedOrders(): Observable<{ success: boolean; message: string; deliveriesCreated: number }> {
        return this.http.post<{ success: boolean; message: string; deliveriesCreated: number }>(
            `${this.apiUrl}/sync-confirmed-orders`,
            {}
        );
    }
}

